x/storage/keeper/keeper.go

```diff
@@ -1738,9 +1739,46 @@ func (k Keeper) garbageCollectionForResource(ctx sdk.Context, deleteStalePolicie
 	return deletedTotal, true
 }
 
-func (k Keeper) MigrationBucket(ctx sdk.Context, srcSP, dstSP *sptypes.StorageProvider, bucketInfo *types.BucketInfo) error {
+func (k Keeper) MigrateBucket(ctx sdk.Context, operator sdk.AccAddress, bucketName string, dstPrimarySPID uint32, dstPrimarySPApproval *common.Approval, approvalBytes []byte) error {
 	store := ctx.KVStore(k.storeKey)
 
+	bucketInfo, found := k.GetBucketInfo(ctx, bucketName)
+	if !found {
+		return types.ErrNoSuchBucket
+	}
+
+	if !operator.Equals(sdk.MustAccAddressFromHex(bucketInfo.Owner)) {
+		return types.ErrAccessDenied.Wrap("Only bucket owner can migrate bucket.")
+	}
+
+	if bucketInfo.BucketStatus == types.BUCKET_STATUS_MIGRATING {
+		return types.ErrInvalidBucketStatus.Wrapf("The bucket already been migrating")
+	}
+
+	srcSP, found := k.spKeeper.GetStorageProvider(ctx, bucketInfo.PrimarySpId)
+	if !found {
+		return sptypes.ErrStorageProviderNotFound.Wrapf("src sp not found")
+	}
+
+	dstSP, found := k.spKeeper.GetStorageProvider(ctx, dstPrimarySPID)
+	if !found {
+		return sptypes.ErrStorageProviderNotFound.Wrapf("dst sp not found")
+	}
+
+	if srcSP.Status != sptypes.STATUS_IN_SERVICE || dstSP.Status != sptypes.STATUS_IN_SERVICE {
+		return sptypes.ErrStorageProviderNotInService.Wrapf(
+			"origin SP status: %s, dst SP status: %s", srcSP.Status.String(), dstSP.Status.String())
+	}
+
+	// check approval
+	if dstPrimarySPApproval.ExpiredHeight < (uint64)(ctx.BlockHeight()) {
```

Comment:
dstPrimarySPApproval.ExpiredHeight `>` (uint64)(ctx.BlockHeight()).

Reply:
Fixed

Reply:
Sorry, I misunderstood here; it should be `<`.

Reply:
Yes, I saw it. The e2e test case did not pass.